mono/mini/mini-posix.c

```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
-		perror ("setflags");
-		return 0;
+	struct timespec ts;
+
+#ifdef PLATFORM_MACOSX
+	kern_return_t ret;
+	clock_id_t mach_clock = SYSTEM_CLOCK;
+	clock_serv_t clock_service;
+	mach_timespec_t mach_ts;
+
+	if ((ret = host_get_clock_service (mach_host_self (), mach_clock, &clock_service)) != KERN_SUCCESS)
+		g_error ("%s: host_get_clock_service () returned %d", __func__, ret);
+
+	if ((ret = clock_get_time (clock_service, &mach_ts)) != KERN_SUCCESS)
+		g_error ("%s: clock_get_time () returned %d", __func__, ret);
+
+	if ((ret = mach_port_deallocate (mach_task_self (), clock_service)) != KERN_SUCCESS)
+		g_error ("%s: mach_port_deallocate () returned %d", __func__, ret);
+
+	ts.tv_sec = mach_ts.tv_sec;
+	ts.tv_nsec = mach_ts.tv_nsec;
+#else
+	clockid_t posix_clock;
+
+	switch (mode) {
+	case MONO_PROFILER_STAT_MODE_PROCESS:
+#ifdef HAVE_CLOCK_NANOSLEEP
+		/*
+		 * If we don't have clock_nanosleep (), measuring the process time
+		 * makes very little sense as we can only use nanosleep () to sleep on
+		 * real time.
+		 */
+		posix_clock = CLOCK_PROCESS_CPUTIME_ID;
+		break;
+#endif
+	case MONO_PROFILER_STAT_MODE_REAL: posix_clock = CLOCK_MONOTONIC; break;
+	default: g_assert_not_reached (); break;
 	}
-	return 1;
-}
+
+	if (clock_gettime (posix_clock, &ts) == -1)
+		g_error ("%s: clock_gettime () returned -1, errno = %d", __func__, errno);
 #endif
 
-void
-mono_runtime_shutdown_stat_profiler (void)
+	return ((gint64) ts.tv_sec * 1000000000) + (gint64) ts.tv_nsec;
+}
+
+static void
+clock_sleep_ns_abs (MonoProfileSamplingMode mode, guint64 ns_abs)
 {
-#ifdef HAVE_LINUX_RTC_H
-	if (rtc_fd >= 0)
-		enable_rtc_timer (FALSE);
+	MONO_PREPARE_BLOCKING;
+
+#ifdef PLATFORM_MACOSX
+	kern_return_t ret;
+	clock_id_t mach_clock = SYSTEM_CLOCK;
+	clock_serv_t clock_service;
+	mach_timespec_t then, remain_unused;
+
+	if ((ret = host_get_clock_service (mach_host_self (), mach_clock, &clock_service)) != KERN_SUCCESS)
+		g_error ("%s: host_get_clock_service () returned %d", __func__, ret);
+
+	then.tv_sec = ns_abs / 1000000000;
+	then.tv_nsec = ns_abs % 1000000000;
+
+	do {
+		ret = clock_sleep (clock_service, TIME_ABSOLUTE, then, &remain_unused);
+
+		if (ret != KERN_SUCCESS && ret != KERN_ABORTED)
+			g_error ("%s: clock_sleep () returned %d", __func__, ret);
+	} while (ret == KERN_ABORTED);
+
+	if ((ret = mach_port_deallocate (mach_task_self (), clock_service)) != KERN_SUCCESS)
+		g_error ("%s: mach_port_deallocate () returned %d", __func__, ret);
+#elif defined (HAVE_CLOCK_NANOSLEEP)
+	int ret;
+	clockid_t posix_clock;
+	struct timespec then;
+
+	switch (mode) {
+	case MONO_PROFILER_STAT_MODE_PROCESS: posix_clock = CLOCK_PROCESS_CPUTIME_ID; break;
+	case MONO_PROFILER_STAT_MODE_REAL: posix_clock = CLOCK_MONOTONIC; break;
+	default: g_assert_not_reached (); break;
+	}
+
+	then.tv_sec = ns_abs / 1000000000;
+	then.tv_nsec = ns_abs % 1000000000;
+
+	do {
+		ret = clock_nanosleep (posix_clock, TIMER_ABSTIME, &then, NULL);
```

Comment:
We don't need to use the `remain` argument. It actually won't even be set since we're passing `TIMER_ABSTIME`. `clock_nanosleep` will simply return immediately if `then` is in the past.